presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.DependencyExtractor;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+
+import java.util.Iterator;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.iterative.rule.Util.transpose;
+
+public class SwapAdjacentWindowsByPartitionsOrder
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof WindowNode)) {
+            return Optional.empty();
+        }
+
+        WindowNode parent = (WindowNode) node;
+
+        PlanNode child = lookup.resolve(parent.getSource());
+        if (!(child instanceof WindowNode)) {
+            return Optional.empty();
+        }
+
+        if ((compare(parent, (WindowNode) child) < 0) && (!dependsOn(parent, (WindowNode) child))) {
+            return Optional.of(transpose(parent, child));
+        }
+        else {
+            return Optional.empty();
+        }
+    }
+
+    private static boolean dependsOn(WindowNode parent, WindowNode child)
+    {
+        return parent.getPartitionBy().stream().anyMatch(symbol -> child.getCreatedSymbols().contains(symbol))
+                || parent.getOrderBy().stream().anyMatch(symbol -> child.getCreatedSymbols().contains(symbol))
+                || parent.getWindowFunctions().values().stream()
+                .map(WindowNode.Function::getFunctionCall)
+                .flatMap(functionCall -> functionCall.getArguments().stream())
```

Comment:
Why not call DependencyExtractor.extractUnique on the FunctionCall directly (vs exploding the arguments first)?